File: src/tauri/commands/backlog.ts
import { safeInvoke } from "./_base";

export type BacklogIssueType = {
  id: number;
  projectId: number;
  name: string;
  color?: string;
  displayOrder?: number;
};

export type BacklogStatus = {
  id: number;
  projectId?: number;
  name: string;
  color?: string;
  displayOrder?: number;
};

export type BacklogCategory = {
  id: number;
  name: string;
  displayOrder?: number;
};

export type BacklogUser = {
  id: number;
  userId?: string;
  name: string;
  roleType?: number;
  lang?: string;
  mailAddress?: string;
  keyword?: string;
};

export function backlogListIssueTypes(projectKey: string) {
  return safeInvoke<BacklogIssueType[]>("backlog_list_issue_types", { projectKey });
}

export function backlogListStatuses(projectKey: string) {
  return safeInvoke<BacklogStatus[]>("backlog_list_statuses", { projectKey });
}

export function backlogListCategories(projectKey: string) {
  return safeInvoke<BacklogCategory[]>("backlog_list_categories", { projectKey });
}

export function backlogListProjectUsers(projectKey: string) {
  return safeInvoke<BacklogUser[]>("backlog_list_project_users", { projectKey });
}

export type BacklogPriority = {
  id: number;
  name: string;
};

export type BacklogIssue = {
  id: number;
  projectId: number;
  issueKey: string;
  keyId: number;
  issueType: BacklogIssueType | null;
  summary: string;
  description: string | null;
  priority: BacklogPriority | null;
  status: BacklogStatus | null;
  assignee: BacklogUser | null;
  category: BacklogCategory[] | null;
  milestone: unknown[] | null;
  startDate: string | null;
  dueDate: string | null;
  estimatedHours: number | null;
  actualHours: number | null;
  parentIssueId: number | null;
  createdUser: BacklogUser | null;
  created: string | null;
  updated: string | null;
};

export type BacklogIssueQuery = {
  projectKey: string;
  count?: number;
  offset?: number;
  statusIds?: number[];
  assigneeIds?: number[];
  issueTypeIds?: number[];
  categoryIds?: number[];
  milestoneIds?: number[];
  priorityIds?: number[];
  keyword?: string;
  sort?: string;
  order?: string;
  parentIssueId?: number;
};

export type BacklogIssueList = {
  issues: BacklogIssue[];
  totalCount: number;
};

export type BacklogProjectLookup = {
  projectId: string;
  projectKey: string;
  projectName: string;
};

export function backlogListIssues(query: BacklogIssueQuery) {
  return safeInvoke<BacklogIssueList>("backlog_list_issues", { query });
}

export function backlogGetProjectLookup(projectKey: string) {
  return safeInvoke<BacklogProjectLookup>("backlog_get_project_lookup", { projectKey });
}

export function backlogListPriorities() {
  return safeInvoke<BacklogPriority[]>("backlog_list_priorities");
}

export type BacklogCreateIssueRequest = {
  projectId: number;
  summary: string;
  issueTypeId: number;
  priorityId: number;
  description?: string;
  assigneeId?: number;
  startDate?: string;
  dueDate?: string;
  estimatedHours?: number;
  actualHours?: number;
  categoryId?: number[];
  milestoneId?: number[];
  parentIssueId?: number;
};

export function backlogCreateIssue(request: BacklogCreateIssueRequest) {
  return safeInvoke<BacklogIssue>("backlog_create_issue", { request });
}

File: src-tauri/src/models/backlog.rs
//! Các kiểu dữ liệu (model) cho Backlog API responses.

use serde::{Deserialize, Serialize};
/// Thông tin dự án từ Backlog API.
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BacklogProject {
    pub id: i64,
    pub project_key: String,
    pub name: String,
    pub archived: Option<bool>,
    pub text_formatting_rule: Option<String>,
    pub display_order: Option<i64>,
}

/// Loại issue trong dự án Backlog.
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BacklogIssueType {
    pub id: i64,
    pub project_id: i64,
    pub name: String,
    pub color: Option<String>,
    pub display_order: Option<i64>,
}

/// Trạng thái issue trong dự án Backlog.
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BacklogStatus {
    pub id: i64,
    pub project_id: Option<i64>,
    pub name: String,
    pub color: Option<String>,
    pub display_order: Option<i64>,
}

/// Danh mục (category) trong dự án Backlog.
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BacklogCategory {
    pub id: i64,
    pub name: String,
    pub display_order: Option<i64>,
}

/// Thông tin user trên Backlog.
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BacklogUser {
    pub id: i64,
    pub user_id: Option<String>,
    pub name: String,
    pub role_type: Option<i64>,
    pub lang: Option<String>,
    pub mail_address: Option<String>,
    pub keyword: Option<String>,
}

/// Priority của issue.
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BacklogPriority {
    pub id: i64,
    pub name: String,
}

/// Milestone / version trong dự án.
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BacklogMilestone {
    pub id: i64,
    pub project_id: i64,
    pub name: String,
    pub description: Option<String>,
    pub start_date: Option<String>,
    pub release_due_date: Option<String>,
    pub archived: Option<bool>,
    pub display_order: Option<i64>,
}

/// Thông tin issue từ Backlog API.
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BacklogIssue {
    pub id: i64,
    pub project_id: i64,
    pub issue_key: String,
    pub key_id: i64,
    pub issue_type: Option<BacklogIssueType>,
    pub summary: String,
    pub description: Option<String>,
    pub priority: Option<BacklogPriority>,
    pub status: Option<BacklogStatus>,
    pub assignee: Option<BacklogUser>,
    pub category: Option<Vec<BacklogCategory>>,
    pub milestone: Option<Vec<BacklogMilestone>>,
    pub start_date: Option<String>,
    pub due_date: Option<String>,
    pub estimated_hours: Option<f64>,
    pub actual_hours: Option<f64>,
    pub parent_issue_id: Option<i64>,
    pub created_user: Option<BacklogUser>,
    pub created: Option<String>,
    pub updated: Option<String>,
}

/// Tham số lọc khi lấy danh sách issues.
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BacklogIssueQuery {
    pub project_key: String,
    #[serde(default = "default_count")]
    pub count: i32,
    #[serde(default)]
    pub offset: i32,
    pub status_ids: Option<Vec<i64>>,
    pub assignee_ids: Option<Vec<i64>>,
    pub issue_type_ids: Option<Vec<i64>>,
    pub category_ids: Option<Vec<i64>>,
    pub milestone_ids: Option<Vec<i64>>,
    pub priority_ids: Option<Vec<i64>>,
    pub keyword: Option<String>,
    pub sort: Option<String>,
    pub order: Option<String>,
    pub parent_issue_id: Option<i64>,
}

fn default_count() -> i32 {
    100
}

/// Kết quả đếm issues.
#[derive(Debug, Serialize, Deserialize)]
pub struct BacklogIssueCount {
    pub count: i64,
}

/// Danh sách issues kèm tổng số.
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BacklogIssueList {
    pub issues: Vec<BacklogIssue>,
    pub total_count: i64,
}

/// Kết quả tra cứu dự án từ Backlog API theo project key.
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BacklogProjectLookup {
    pub project_id: String,
    pub project_key: String,
    pub project_name: String,
}
